### PR TITLE
feat: add MOCHA_OPTIONS env variable

### LINKED ISSUE #4330: 🚀 Feature: Support setting options via environment vars
As discussed in #4232, it can be difficult to pass mocha command-line options through `npm run` scripts.  As @boneskull said:

> yargs has a feature that supports setting options via an environment variable; we should probably take advantage of that at some point. Might even be an issue for it already.

I didn't see any open issues for this so figured I would open one.

### PR DESCRIPTION
**Pull request by icholy**
### Description of the Change

The `MOCHA_OPTIONS` environment variable is a space-separated list of command-line options. It essentially copies node's approach with [NODE_OPTIONS](https://nodejs.org/api/cli.html#node_optionsoptions)

The config priority is:

1. Command-line args
2. `MOCHA_OPTIONS` environment variable.
3. RC file (`.mocharc.c?js`, `.mocharc.ya?ml`, `mocharc.json`)
4. `mocha` prop of `package.json`
5. default configuration (`lib/mocharc.json`)

### Alternate Designs

Yargs has support for env variables: https://yargs.js.org/docs/#api-reference-envprefix
However, using this feature would result in a separate env variable for each option:

Ex:

* `MOCHA_BAIL`
* `MOCHA_ALLOW_UNCAUGHT`
* `MOCHA_FAIL_ZERO`
* etc ...

I decided to to go with `MOCHA_OPTIONS` because it seems less intrusive and that's how node does it too.

### Why should this be in core?

I don't see how it could live anywhere else.

### Benefits

My specific use-case is specifying `MOCHA_OPTIONS='--inspect-brk'` when calling `npm run test` from vim.

### Possible Drawbacks

* More code is always a liability.
* More config sources can make things more difficult to debug.

### Applicable issues

Fixes #4330.

### CODE REVIEW

**Comment by icholy** [comment]
@juergba have you had a chance to look this over?

**Comment by icholy** [comment]
@outsideris I still need this feature. Would I be able to get some feedback?

**Comment by JoshuaKGoldberg** [comment]
👋 coming back to this @icholy, are you still interested in working on this PR? As of https://github.com/mochajs/mocha/issues/5027 there are maintainers who can review it now. No worries if you no longer have time - but if you do that'd be great!

**Comment by icholy** [comment]
Yes!

**Comment by JoshuaKGoldberg** [comment]
Yay! We're not quite ready to start reviewing PRs yet, but hope to soon. 

Note that I don't see any linked issues. The contributing guidelines right now ask for that so that folks can discuss the change first. We'll continue to require that after https://github.com/mochajs/mocha/pull/5038. Was there any impetus for this PR besides what's in the description?

**Comment by icholy** [comment]
There's no separate issue, the PR description contains all context I have. I can create an issue if you'd like though.

**Comment by JoshuaKGoldberg** [comment]
That would be lovely, yes please!

Note that the three of us new maintainers are still ramping up on the project, so it might be some weeks (or even months 😞) before we have enough context to really triage the issue and review this PR. We'll do our best to get to it soon though!

**Comment by icholy** [comment]
That's alright. It's been two years, what's another month or two?

File: lib/cli/options.js
Comment (JoshuaKGoldberg) [inline lib/cli/options.js:1]: [Docs] The new `MOCHA_OPTIONS` env variable will need to be documented. https://mochajs.org/#merging is the main place for configuration/options merging info that I know of. There might be others (someone yell at me if there are?).

File: lib/cli/options.js
Comment (JoshuaKGoldberg) [inline lib/cli/options.js:255]: [Praise] This is really cleanly implemented. Nicely done! ❤️

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Sweet! Really pleasing to see such a clean implementation. And with thorough tests! 👏 

Just requesting changes on docs.

File: lib/cli/options.js
Comment (icholy) [inline lib/cli/options.js:214]: @JoshuaKGoldberg I'm not sure about the priority ordering here. I can't decide if env vars should be higher up or not.

Comment (shadowspawn) [reply lib/cli/options.js:214]: (Drive-by comment from a Yargs and Commander contributor, but not a Mocha user so do not have domain insights.)

Environment variables can be specified on the command-line, and I think are often treated as second in priority behind the command-line args.

Yargs specifies:
> Program arguments are defined in this order of precedence:
> 1. Command line args
> 2. Env vars
> 3. Config file/objects
> 4. Configured defaults

https://yargs.js.org/docs/#api-reference-envprefix

Comment (icholy) [reply lib/cli/options.js:214]: @shadowspawn that's my gut feeling too after playing with it for a while. @JoshuaKGoldberg cool if I update the PR?

Comment (JoshuaKGoldberg) [reply lib/cli/options.js:214]: Sounds good to me 🙂 thanks for the insights @shadowspawn!

File: lib/cli/options.js
Comment (icholy) [reply lib/cli/options.js:1]: I've updated the docs.

**Comment by icholy** [comment]
@JoshuaKGoldberg do you need anything more from my end for this PR?

File: docs/index.md
Comment (JoshuaKGoldberg) [inline docs/index.md:2219]: [Docs] I don't think we can expect all users to intuitively know how this setting CLI args through an env variable looks. How about a bit of explicit help like...

````suggestion
The `MOCHA_OPTIONS` environment variable may be used to specify command line arguments. These arguments take priority over those found in configuration files.

For example, setting the `bail` and `retries` options:

```bash
$ MOCHA_OPTIONS="--bail --retries 3" mocha
```
````

**Review by JoshuaKGoldberg — APPROVED** [review]
This looks great to me, thanks ✨ !. Just the one docs suggestion.

We should get a review by another member of @mochajs/maintenance-crew too, just to be safe.

**Comment by JoshuaKGoldberg** [comment]
Btw I generally wait to be [explicitly re-requested review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#re-requesting-a-review) before taking a look - though pinging the way you did works too. That's why this took so long. Sorry for keeping you. 🙂

File: docs/index.md
Comment (icholy) [reply docs/index.md:2219]: Good call.

**Comment by icholy** [comment]
cc @voxpelli for review

**Comment by JoshuaKGoldberg** [comment]
Sad to say nothing 😞 we're just waiting on the other maintainers. I know they've been pretty busy this season but I've sent pings privately.

**Review by voxpelli — APPROVED** [review]
> I don't see how it could live anywhere else.

One can set up a `.mocharc.js` file that reads from `process.env` as exemplified here: https://github.com/mochajs/mocha/blob/master/example/config/.mocharc.js#L3-L5

But after thinking about this I see little harm in adding this option.

The one thing that we maybe would want to add down the line is a flag that opts out from `MOCHA_OPTIONS` to be read and used, but lets deal with that when the need is proven.

Approved 👍

(Sorry for slow review, have been crunching on [neostandard](https://github.com/neostandard/neostandard) + been pondering whether an environment option like this could become an injection attack of some kind, but after some googling I can't find much written that would discourage it)

**Comment by icholy** [comment]
> been pondering whether an environment option like this could become an injection attack of some kind

If someone has write access to your env vars its already game over.

**Comment by voxpelli** [comment]
> > been pondering whether an environment option like this could become an injection attack of some kind
> 
> If someone has write access to your env vars its already game over.

Yeah, but sometimes people try to issue vulnerability reports for the strangest of things

**Comment by JoshuaKGoldberg** [comment]
This was released in [`mocha@10.5.0`](https://www.npmjs.com/package/mocha/v/10.5.0). 🚀

**Pull request merged**
